test/features/test_organized_edge_detection.cpp

```diff
@@ -0,0 +1,160 @@
+          if (col == left_col - 1 || col == right_col) {
+            large_inner_perimeter_.indices.push_back(row * organized_test_cloud->width +
+                                                     col);
+          }
+        }
+
+        organized_test_cloud->at(col, row).x = x * SYNTHETIC_CLOUD_RESOLUTION;
+        organized_test_cloud->at(col, row).y = y * SYNTHETIC_CLOUD_RESOLUTION;
+        organized_test_cloud->at(col, row).z = depth;
+      }
+    }
+
+    return organized_test_cloud;
+  }
+};
+} // namespace
+
+//////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+/*
+This test is designed to ensure that the organized edge detection appropriately
+classifies occluding and occluding edges and to specifically detect the type of
+regression detailed in PR 4275 (https://github.com/PointCloudLibrary/pcl/pull/4275).
+This test works by generating a synthetic cloud of one square slightly in front of
+another square, so that occluding edges and occluded edges are generated.  The
+regression introduced in PCL 1.10.1 was a logic bug that caused both occluding and
+occluded edges to be miscategorized as occluding edges.  This test should catch
+this and similar bugs.
+*/
+TEST_F(OrganizedPlaneDetectionTestFixture, OccludedAndOccludingEdges)
+{
+  const auto MAX_SEARCH_NEIGHBORS = 8;
+
+  // The depth discontinuity check to determine whether an edge exists is linearly
+  // dependent on the depth of the points in the cloud (not a fixed distance).  The
+  // algorithm iterates through each point in the cloud and finding the neighboring
+  // point with largest discontinuity value.  That value is compared against a threshold
+  // multiplied by the actual depth value of the point. Therefore:
+  // abs(SYNTHETIC_CLOUD_DEPTH_DISCONTINUITY) must be greater than
+  // DEPTH_DISCONTINUITY_THRESHOLD * abs(SYNTHETIC_CLOUD_BASE_DEPTH)
+  const auto DEPTH_DISCONTINUITY_THRESHOLD =
+      SYNTHETIC_CLOUD_DEPTH_DISCONTINUITY / (SYNTHETIC_CLOUD_BASE_DEPTH * 1.1f);
+
+  auto oed = pcl::OrganizedEdgeBase<pcl::PointXYZ, pcl::Label>();
+  auto labels = pcl::PointCloud<pcl::Label>();
+  auto label_indices = std::vector<pcl::PointIndices>();
+
+  oed.setEdgeType(oed.EDGELABEL_OCCLUDING | oed.EDGELABEL_OCCLUDED);
+  oed.setInputCloud(cloud_);
+  oed.setDepthDisconThreshold(DEPTH_DISCONTINUITY_THRESHOLD);
+  oed.setMaxSearchNeighbors(MAX_SEARCH_NEIGHBORS);
+  oed.compute(labels, label_indices);
+
+  EXPECT_EQ(label_indices[1].indices.size(), small_outer_perimeter_.indices.size());
+  for (auto i = 0; i < small_outer_perimeter_.indices.size(); ++i) {
+    EXPECT_EQ(label_indices[1].indices[i], small_outer_perimeter_.indices[i]);
+  }
+
+  EXPECT_EQ(label_indices[2].indices.size(), large_inner_perimeter_.indices.size());
+  for (auto i = 0; i < large_inner_perimeter_.indices.size(); ++i) {
+    EXPECT_EQ(label_indices[2].indices[i], large_inner_perimeter_.indices[i]);
+  }
```

Comment:
These checks rely on the order of traversal of the algorithm and they shouldn't. Hence my original suggestions to use a set to store the expected results and remove from there. The important idea is that you can't rely on the returned order in any way.